go/vt/vtgate/balancer/session_test.go

```diff
@@ -0,0 +1,342 @@
+	}
+
+	tablets := []*discovery.TabletHealth{
+		{
+			Tablet: &topodatapb.Tablet{
+				Alias: &topodatapb.TabletAlias{
+					Cell: "local",
+					Uid:  100,
+				},
+				Keyspace: "keyspace",
+				Shard:    "0",
+			},
+			Target: &querypb.Target{
+				Keyspace:   "keyspace",
+				Shard:      "0",
+				TabletType: topodatapb.TabletType_REPLICA,
+				Cell:       "local",
+			},
+			Serving: true,
+		},
+
+		{
+			Tablet: &topodatapb.Tablet{
+				Alias: &topodatapb.TabletAlias{
+					Cell: "local",
+					Uid:  101,
+				},
+				Keyspace: "keyspace",
+				Shard:    "0",
+			},
+			Target: &querypb.Target{
+				Keyspace:   "keyspace",
+				Shard:      "0",
+				TabletType: topodatapb.TabletType_REPLICA,
+				Cell:       "local",
+			},
+			Serving: true,
+		},
+	}
+
+	// Pick for a specific session UUID
+	picked1 := b.Pick(target, tablets, WithSessionUUID("a"))
+	require.NotNil(t, picked1)
+
+	// Pick again with same session hash, should return same tablet
+	picked2 := b.Pick(target, tablets, WithSessionUUID("a"))
+	require.Equal(t, picked1, picked2, fmt.Sprintf("expected %s, got %s", tabletAlias(picked1), tabletAlias(picked2)))
+
+	// Pick with different session hash, empirically know that it should return tablet2
+	picked3 := b.Pick(target, tablets, WithSessionUUID("b"))
+	require.NotNil(t, picked3)
+	require.NotEqual(t, picked2, picked3, fmt.Sprintf("expected different tablets, got %s for both", tabletAlias(picked3)))
+}
+
+func TestPickPreferLocal(t *testing.T) {
+	b := createSessionBalancer(t)
+
+	target := &querypb.Target{
+		Keyspace:   "keyspace",
+		Shard:      "0",
+		TabletType: topodatapb.TabletType_REPLICA,
+		Cell:       "local",
+	}
+
+	tablets := []*discovery.TabletHealth{
+		{
+			Tablet: &topodatapb.Tablet{
+				Alias: &topodatapb.TabletAlias{
+					Cell: "local",
+					Uid:  100,
+				},
+				Keyspace: "keyspace",
+				Shard:    "0",
+			},
+			Target: &querypb.Target{
+				Keyspace:   "keyspace",
+				Shard:      "0",
+				TabletType: topodatapb.TabletType_REPLICA,
+				Cell:       "local",
+			},
+			Serving: true,
+		},
+
+		{
+			Tablet: &topodatapb.Tablet{
+				Alias: &topodatapb.TabletAlias{
+					Cell: "local",
+					Uid:  101,
+				},
+				Keyspace: "keyspace",
+				Shard:    "0",
+			},
+			Target: &querypb.Target{
+				Keyspace:   "keyspace",
+				Shard:      "0",
+				TabletType: topodatapb.TabletType_REPLICA,
+				Cell:       "local",
+			},
+			Serving: true,
+		},
+
+		{
+			Tablet: &topodatapb.Tablet{
+				Alias: &topodatapb.TabletAlias{
+					Cell: "external",
+					Uid:  200,
+				},
+				Keyspace: "keyspace",
+				Shard:    "0",
+			},
+			Target: &querypb.Target{
+				Keyspace:   "keyspace",
+				Shard:      "0",
+				TabletType: topodatapb.TabletType_REPLICA,
+				Cell:       "external",
+			},
+			Serving: true,
+		},
+	}
+
+	// Pick should prefer local cell
+	picked1 := b.Pick(target, tablets, WithSessionUUID("a"))
+	require.NotNil(t, picked1)
+	require.Equal(t, "local", picked1.Target.Cell)
```

Comment:
It's probably worth doing this in a loop 10 times or something, no? To better test the preference.